Replaced angular.extend with Utilities.extend

bjarnef opened the pull request:
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes https://github.com/umbraco/Umbraco-CMS/issues/7718

### Description
This PR replaced of references on `angular.extend()` with `Utilities.extend()` except in the `Utilities` helper itself.

Note that `angular.extend(dst, src)` could be use with two or more arguments to pass in multiple source objects.
This was used a few places like `var object = angular.extend({}, object1, object2);` - we could implement this, but not sure if it is worth if as it was only use a few places and if AngularJS is replaced later or we later want to use vanilla JS here.

For example `var opts = Utilities.extend({}, options, scope.umbAceEditor);` would returns an empty object, while `var opts = Utilities.extend(options, scope.umbAceEditor);` returns the same as the previous `var opts = angular.extend({}, options, scope.umbAceEditor);`.

bjarnef commented:
Maybe this is one you would like to review @nathanwoulfe and I think the original issue then can be closed 😁🔥🚀

ronaldbarendse commented:
> For example `var opts = Utilities.extend({}, options, scope.umbAceEditor);` would returns an empty object, while `var opts = Utilities.extend(options, scope.umbAceEditor);` returns the same as the previous `var opts = angular.extend({}, options, scope.umbAceEditor);`.

If that's the case, it's probably a bug in `Utilities.extend`, as similar methods all update the first object that's passed in and also returns that same object:
- https://docs.angularjs.org/api/ng/function/angular.extend
- https://api.jquery.com/jquery.extend/
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

ronaldbarendse commented:
Looks like `Utilities.extend` is a facade to `angular.extend` that only supports 2 parameters:

https://github.com/umbraco/Umbraco-CMS/blob/5e4d326ec2927d9895154afdbb38c493059211c9/src/Umbraco.Web.UI.Client/src/utilities.js#L33-L37

So that's the problem with using `Utilities.extend({}, options, scope.umbAceEditor);` 😉

bjarnef commented on src/Umbraco.Web.UI.Client/src/common/services/keyboard.service.js:
```diff
@@ -264,7 +264,7 @@ function keyboardService($window, $timeout) {
         var elt;
         // Initialize opt object
-        opt   = angular.extend({}, defaultOpt, opt);
+        opt   = Utilities.extend(defaultOpt, opt);
```
@ronaldbarendse the `angular.extend({}, object1, object2)` allowed a second it more parameters (source objects) to merged with the empty object (however note angular.extend does not support recursive merge (deep copy)): angular.extend does not support recursive merge (deep copy).

So if we just used `Utilities.extend({}, defaultOpt, opt);` it would return an empty object, because the `Utilities.extend` only use `angular.extend` with two arguments.

ronaldbarendse replied:
This still needs to be updated to `opt = Utilities.extend({}, defaultOpt, opt);`, as it would otherwise merge the current options (`opt`) into the default options (`defaultOpt`) and cause lots of issues when used multiple times...

bjarnef replied:
I recall I just logged the `opt` variable before and after here, but not sure if it would have other side effects as you mention.

ronaldbarendse replied:
After merging this PR, it's now fixed in https://github.com/umbraco/Umbraco-CMS/pull/10348 👍🏻

bjarnef commented:
@ronaldbarendse yes, not sure if we wanted to support the format `Utilities.extend({}, options, scope.umbAceEditor);` since we only used that format a few places in code and if we later swich to another framework or use vanilla JS.

nathanwoulfe commented:
Pretty sure boring old Object.assign() allows multiples sources, so `var obj = Object.assign({}, foo, bar) would assign props from foo and bar to the empty object, and return that new object. 

Given that, do we need to abstract anything? Could we just use Object.assign instead?

bjarnef commented:
@nathanwoulfe it seems `Object.assign()` doesn't by default do a deep clone https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#deep_clone which `angular.extend()` didn't support.

Do you know how we can change the `Utilities.extend()` to support the `angular.extend({}, src1, src2)` format while still working with  `angular.extend(dst, src)` .. or if we just can handle this with ``Object.assign()` instead?

markadrake commented:
@bjarnef @nathanwoulfe I'm definitely not an angular expert - just wondering what the differences are between `angular.extend` and `Object.assign`?

https://codesandbox.io/s/polished-grass-61iqt

nathanwoulfe commented:
At some point all the angular js facade stuff will disappear, and we'll be using Object.assign everywhere - I'm not sure of use cases where we'd want to be deep-copying (I haven't had a good look, but most cases are extending config objects with additional fields, so Object.assign would be perfectly ok).

Biggest difference between the two methods is .extend is angular-friendly, and preserves the $$hashKey property, which is critical when extending angular-managed objects. If none of our extends are called on angular objects, then that's not an issue.

nathanwoulfe commented:
And extending our extend in Utilities is trivial, just needs the spread operator on the src param:

![image](https://user-images.githubusercontent.com/3248070/112067270-80b9d000-8bb3-11eb-86fc-d4bdb3b539fd.png)

bjarnef commented:
@nathanwoulfe I have updated it to use the spread operator.
However I get a different in this. This first `opts` is the current format of the object.

```
var opts = angular.extend({}, options, scope.umbAceEditor);
console.log("opts 1", opts);

var opts2 = Utilities.extend({}, options, scope.umbAceEditor);
console.log("opts 2", opts2);
```

![image](https://user-images.githubusercontent.com/2919859/112109998-937fd500-8bb2-11eb-8c1b-b09db0c036aa.png)

bjarnef commented:
I tried with and without spread operator in the `Utilities.extend()`, but it didn't seem to make a difference on the output here - probably because we only pass in a single source object.
https://github.com/umbraco/Umbraco-CMS/pull/10023/commits/29b107c02aafecb08ef60258c64ed719508e4aa1

nathanwoulfe commented:
Hmm, that's interesting - my test case was entirely in the browser console, so maybe different result compared to the gulpified JavaScript?

bjarnef commented:
I have reverted back to the original format using an empty destination object `{}` and added the spread operator to the `extend()` function, but currently it doesn't work as the "Chrome" theme isn't applied to ace editor.

![image](https://user-images.githubusercontent.com/2919859/112111192-29682f80-8bb4-11eb-92b8-47ed3a9ba084.png)

bjarnef commented:
Yes, it could be some configuration in Gulp and/or in Babel.
https://github.com/babel/gulp-babel/issues/142#issuecomment-660411880

Maybe @nielslyngsoe or @madsrasmussen can help here?

bjarnef commented:
@nathanwoulfe okay, I found this https://davidwalsh.name/merge-objects
and it seems when configurating this Babel plugin it works: https://babeljs.io/docs/en/babel-plugin-proposal-object-rest-spread

![image](https://user-images.githubusercontent.com/2919859/112122430-8ec21d80-8bc0-11eb-9f26-10050006171e.png)

Fixed in https://github.com/umbraco/Umbraco-CMS/pull/10023/commits/6f3b0f9eb24a3f7d24970da6cd983b953457bf69

bjarnef commented:
I have leaved the `loose` property to the default value `false`.
Setting `useBuiltIns` property will use `Object.assign` directly instead of the Babel's `extends` helper.

bjarnef commented:
@ronaldbarendse @nathanwoulfe are you happy with the latest changes? 😁

bjarnef commented:
@nathanwoulfe could you have another look at this? It would be great to solve the last part of your original issue/feature request ✨

ronaldbarendse commented on src/Umbraco.Web.UI.Client/src/views/propertyeditors/datepicker/datepicker.controller.js:
```diff
@@ -24,7 +24,10 @@ function dateTimePickerController($scope, angularHelper, dateHelper, validationM
 
         // map the user config
-        $scope.model.config = angular.extend(config, $scope.model.config);
+        Utilities.extend(config, $scope.model.config);
+
+        // map back to the model
+        $scope.model.config = config;
```
The `Utilities.extend` function also returns the merged object, so this can be simplified as:
```suggestion
        $scope.model.config = Utilities.extend(config, $scope.model.config);
```

bjarnef commented on src/Umbraco.Web.UI.Client/src/utilities.js:
```diff
@@ -34,7 +34,7 @@
      * Use this with Angular objects, for vanilla JS objects, use Object.assign()
      */
-    const extend = (dst, src) => angular.extend(dst, src);
+    const extend = (dst, ...src) => angular.extend(dst, ...src);
```
@nathanwoulfe do you have some feedback on this? 🤓

nathanwoulfe commented:
@bjarnef @ronaldbarendse it took a while, but we got there - I've gone with the alias in utilities:

`const extend = angular.extend;` 

because it's the lowest impact change. No need to modify parameters, it's a direct alias and will Just Work. No traces of `angular.extend` exist anywhere in the codebase, outside of the Utilities facade. Happy days!

bjarnef commented on src/Umbraco.Web.UI.Client/src/utilities.js:
```diff
@@ -33,8 +33,9 @@
     /**
      * Facade to angular.extend
      * Use this with Angular objects, for vanilla JS objects, use Object.assign()
+     * This is an alias as it to allow passing an unknown number of arguments
```
@nathanwoulfe a minor issue with the wording here "as it to allow".

nathanwoulfe replied:
That's what happens when I type before coffee. Will fix that up, and I need to double check the changes in keyboard.service.js, I think there's a bit in there that needs to change back to the original, else we're updating the default options, which we don't want.

nathanwoulfe replied:
@bjarnef would you mind pushing a new PR from your feature branch, to fix that typo and update keyboard.service.js:

ln26:         opt   = Utilities.extend({}, defaultOpt, opt);

I can't push to the branch anymore, probably because the original PR has been merged... If you could make that change, I'll merge it straight away.

ronaldbarendse commented:
@nathanwoulfe I see you've also merged the changes to `.babelrc` and `package.json`, which are not needed anymore when using the function alias: probably best to revert these...

ronaldbarendse commented:
@bjarnef The spread operator isn't used anymore in this PR, as it now uses the function alias, right? Lets not try to add features/NPM packages that aren't needed to keep everything clean (and not introduce side-effects because of the additional transpiling) 😉
